L1Trigger/TrackFindingTracklet/src/DR.cc

```diff
@@ -0,0 +1,163 @@
+      vector<Track*>& input = input_[channel];
+      const int streamTrackId = offsetTrack + channel;
+      const int offsetStub = streamTrackId * setup_->numLayers();
+      const StreamTrack streamTrack = streamsTrack[streamTrackId];
+      for (int frame = 0; frame < (int)streamTrack.size(); frame++) {
+        const FrameTrack& frameTrack = streamTrack[frame];
+        if (frameTrack.first.isNull()) {
+          input.push_back(nullptr);
+          continue;
+        }
+        vector<Stub*> stubs;
+        stubs.reserve(setup_->numLayers());
+        for (int layer = 0; layer < setup_->numLayers(); layer++) {
+          const FrameStub& frameStub = streamsStub[offsetStub + layer][frame];
+          if (frameStub.first.isNull())
+            continue;
+          TTBV ttBV = frameStub.second;
+          ttBV >>= dataFormats_->format(Variable::z, Process::kfin).width();
+          ttBV >>= dataFormats_->format(Variable::phi, Process::kfin).width();
+          ttBV >>= dataFormats_->format(Variable::r, Process::kfin).width();
+          const TTBV stubId(ttBV, channelAssignment_->widthStubId(), 0, true);
+          ttBV >>= channelAssignment_->widthStubId();
+          const TTBV layerId(ttBV, channelAssignment_->widthLayerId(), 0);
+          stubs_.emplace_back(frameStub, layerId.val(), stubId.val(), layer);
+          stubs.push_back(&stubs_.back());
+        }
+        tracks_.emplace_back(frameTrack, stubs);
+        input.push_back(&tracks_.back());
+      }
+      // remove all gaps between end and last track
+      for (auto it = input.end(); it != input.begin();)
+        it = (*--it) ? input.begin() : input.erase(it);
+    }
+  }
+
+  // fill output products
+  void DR::produce(StreamsStub& accpetedStubs,
+                   StreamsTrack& acceptedTracks,
+                   StreamsStub& lostStubs,
+                   StreamsTrack& lostTracks) {
+    const int offsetTrack = region_ * channelAssignment_->numNodesDR();
+    for (int node = 0; node < channelAssignment_->numNodesDR(); node++) {
+      const int channelTrack = offsetTrack + node;
+      const int offsetStub = channelTrack * setup_->numLayers();
+      // remove duplicated tracks, no merge of stubs, one stub per layer expected
+      vector<Track*> cms(channelAssignment_->drNumComparisonModules(), nullptr);
+      vector<Track*>& tracks = input_[node];
+      for (Track*& track : tracks) {
+        if (!track)
+          // gaps propagate trough chain and appear in output stream
+          continue;
+        for (Track*& trackCM : cms) {
+          if (!trackCM) {
+            // tracks used in CMs propagate trough chain and appear in output stream unaltered
+            trackCM = track;
+            break;
+          }
+          if (equalEnough(track, trackCM)) {
+            // tracks compared in CMs propagate trough chain and appear in output stream as gap if identified as duplicate or unaltered elsewise
```

Comment:
In the f/w tracks are output continuously and no gaps appear in the output, we should discuss the best way for the f/w to match